extensions-core/s3-extensions/src/main/java/io/druid/storage/s3/S3DataSegmentPusher.java

```diff
@@ -163,21 +155,25 @@ public DataSegment push(final File indexFilesDir, final DataSegment inSegment, f
     );
   }
 
-  private void putObject(String bucketName, String path, S3Object object, boolean replaceExisting)
-      throws ServiceException
+  private void uploadFileAndClear(AmazonS3Client s3Client, String bucket, String key, File file, boolean replaceExisting)
   {
-    object.setBucketName(bucketName);
-    object.setKey(path);
+    final PutObjectRequest indexFilePutRequest = new PutObjectRequest(bucket, key, file);
+
     if (!config.getDisableAcl()) {
-      object.setAcl(GSAccessControlList.REST_CANNED_BUCKET_OWNER_FULL_CONTROL);
+      indexFilePutRequest.setAccessControlList(
+          S3Utils.grantFullControlToBucketOwver(s3Client, bucket)
+      );
     }
 
-    log.info("Pushing %s.", object);
-
-    if (!replaceExisting && S3Utils.isObjectInBucket(s3Client, bucketName, object.getKey())) {
-      log.info("Skipping push because key [%s] exists && replaceExisting == false", object.getKey());
+    // TODO: maybe the below check can be moved out to the caller
+    if (!replaceExisting && S3Utils.isObjectInBucketIgnoringPermission(s3Client, bucket, key)) {
+      log.info("Skipping push because key [%s] exists && replaceExisting == false", key);
     } else {
-      s3Client.putObject(bucketName, object);
+      log.info("Pushing [%s] to bucket[%s] and key[%s].", file, bucket, key);
+      s3Client.putObject(indexFilePutRequest);
     }
+
+    log.info("Deleting file [%s]", file.getAbsolutePath());
+    file.delete();
```

Comment:
Do we care if it was successful?

Reply:
Not sure. The original code didn't care about it.